### PR TITLE
[feat]Curveadm: add debug mode

### PR DESCRIPTION
**Pull request by Songjf-ttk**
Related to [#252](https://github.com/opencurve/curveadm/issues/252).

Solution: Add a debug option to the configuration file curvedm.cfg.

Test:
image
The container was not automatically deleted when an error occurred

### CODE REVIEW

**Comment by Wine93** [comment]
Hi @Songjf-ttk, do you think is it more convenient to add a `debug` flag for `format` and `deploy` command? 

e.g.

```bash
curveadm format --debug 
curveadm deploy --debug
```

**Comment by Songjf-ttk** [comment]
>  Yes, I think so. I will modify my code as soon as possible.

File: cli/command/format.go [hunk 1]
```diff
@@ -59,9 +66,10 @@ var (
 )
 
 type formatOptions struct {
-	filename   string
-	showStatus bool
-	stopFormat bool
+	filename    string
+	showStatus  bool
+	stopFormat  bool
+	debugFormat bool
```
Comment (Wine93) [inline cli/command/format.go:72]: ditto.

File: cli/command/deploy.go [hunk 2]
```diff
@@ -238,6 +241,7 @@ func genDeployPlaybook(curveadm *cli.CurveAdm,
 		} else if step == CREATE_LOGICAL_POOL {
 			options[comm.KEY_CREATE_POOL_TYPE] = comm.POOL_TYPE_LOGICAL
 		}
+		options[comm.DEBUG_MODE] = debugDeploy
```
Comment (Wine93) [inline cli/command/deploy.go:244]: Sugesstion:

```
options := map[string]interface{}{
    comm.DEBUG_MODE: debug,
}
if step == CREATE_PHYSICAL_POOL {
    // do something
}
...
```

File: cli/command/format.go [hunk 3]
```diff
@@ -93,25 +102,41 @@ func genFormatPlaybook(curveadm *cli.CurveAdm,
 		return nil, errno.ERR_NO_DISK_FOR_FORMATTING
 	}
 
-	if options.showStatus && options.stopFormat {
+	showStatus := options.showStatus
+	stopFormat := options.stopFormat
+	debugFormat := options.debugFormat
+	if showStatus && stopFormat {
```
Comment (Wine93) [inline cli/command/format.go:108]: Maybe you can do this in `PreRunE` phase, see:
https://github.com/opencurve/curveadm/blob/956c953249a5596ba300bb7541b8c69860284e52/cli/command/deploy.go#L139

File: internal/task/task/bs/format.go [hunk 4]
```diff
@@ -247,6 +249,10 @@ func NewFormatChunkfilePoolTask(curveadm *cli.CurveAdm, fc *configure.FormatConf
 	formatScriptPath := fmt.Sprintf("%s/format.sh", layout.ToolsBinDir)
 	formatCommand := fmt.Sprintf("%s %s %d %d %s %s", formatScriptPath, layout.FormatBinaryPath,
 		usagePercent, DEFAULT_CHUNKFILE_SIZE, layout.ChunkfilePoolDir, layout.ChunkfilePoolMetaPath)
+	debugmode := curveadm.MemStorage().Get(comm.DEBUG_MODE).(bool)
+	if debugmode {
+		glg.Get().SetLevel(glg.DEBG)
+	}
```
Comment (caoxianfei1) [inline internal/task/task/bs/format.go:255]: no need to adjust the log level here. the log level should be determined by the configuration file.

File: internal/task/task/checker/network.go [hunk 5]
```diff
@@ -323,6 +324,12 @@ func NewCleanEnvironmentTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig)
 		return nil, err
 	}
 
+	debugmode := curveadm.MemStorage().Get(comm.DEBUG_MODE).(bool)
+	if debugmode {
+		glg.Get().SetLevel(glg.DEBG)
+		return nil, nil
+	}
```
Comment (caoxianfei1) [inline internal/task/task/checker/network.go:331]: ditto.

File: internal/task/task/common/create_container.go [hunk 6]
```diff
@@ -279,6 +280,7 @@ func NewCreateContainerTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig) (
 		Init:        true,
 		Name:        hostname,
 		Privileged:  true,
+		Remove:      !debugmode,
```
Comment (caoxianfei1) [inline internal/task/task/common/create_container.go:283]: here is a problem, the service container exit but don't delete it forever util delete it manually.

File: internal/task/task/bs/format.go (hunk 4 above)
Comment (Songjf-ttk) [reply internal/task/task/bs/format.go]: There are two ways to implement debug mode. The first is to directly modify curveadm.cfg and read the log level through the configuration file. The second method is to configure the debug mode through the command line, such as :
```shell
curveadm format --debug 
curveadm deploy --debug
```
After communicating with @Wine93 this time, I chose this solution. This Solution set the debug mode in the command line, which may conflict with the log level in curveadm.cfg, so it is necessary to modify the log level.

File: cli/command/format.go [hunk 7]
```diff
@@ -133,6 +170,11 @@ func runFormat(curveadm *cli.CurveAdm, options formatOptions) error {
 	var err error
 	var fcs []*configure.FormatConfig
 	diskRecords := curveadm.DiskRecords()
+	debugmode   := options.debug
```
Comment (caoxianfei1) [inline cli/command/format.go:173]: duplicate with line 128?
debug := options.debug
debugmode := options.debug

**Comment by caoxianfei1** [comment]
@Songjf-ttk look at the other comments.

**Comment by caoxianfei1** [comment]
@Songjf-ttk 
image
The format complete but status is formatting when test it. You may need to lookup `format_status.go` file and modify the logic. 
By the way, should we provide a command such `curveadm format --clean`? In this way we can delete these format containers that exist but exited because we use `curveadm format --debug` for avoid skipping when reformatting.
image

File: internal/task/task/bs/format.go (hunk 4 above)
Comment (Wine93) [reply internal/task/task/bs/format.go]: > There are two ways to implement debug mode. The first is to directly modify curveadm.cfg and read the log level through the configuration file. The second method is to configure the debug mode through the command line, such as :
> 
> ```shell
> curveadm format --debug 
> curveadm deploy --debug
> ```
> 
> After communicating with @Wine93 this time, I chose this solution. This Solution set the debug mode in the command line, which may conflict with the log level in curveadm.cfg, so it is necessary to modify the log level.

yep, maybe you can provide a function for curveadm to set log level.

**Comment by Songjf-ttk** [comment]
> @Songjf-ttk image The format complete but status is formatting when test it. You may need to lookup `format_status.go` file and modify the logic. By the way, should we provide a command such `curveadm format --clean`? In this way we can delete these format containers that exist but exited because we use `curveadm format --debug` for avoid skipping when reformatting. image

1. The reason status is 'formatting' is because curveadm determines the status based on the container's  Status. It has now been fixed:
```
else if len(*s.containerStatus) > 1 && !strings.Contains(*s.containerStatus, "Exited")
```
2. add a command clean to delete the container
```
curveadm format --clean
```

**Comment by caoxianfei1** [comment]
@Songjf-ttk please squash to one commit.

File: cli/command/format.go [hunk 8]
```diff
@@ -56,14 +65,42 @@ var (
...
 
 type formatOptions struct {
-	filename   string
-	showStatus bool
-	stopFormat bool
+	filename    string
+	showStatus  bool
+	stopFormat  bool
+	debug       bool
+	cleanFormat bool
```
Comment (caoxianfei1) [inline cli/command/format.go:79]: clean instead of cleanFormat is better?

File: cli/command/format.go [hunk 9]
```diff
@@ -82,6 +122,8 @@ func NewFormatCommand(curveadm *cli.CurveAdm) *cobra.Command {
 	flags.StringVarP(&options.filename, "formatting", "f", "format.yaml", "Specify the configure file for formatting chunkfile pool")
 	flags.BoolVar(&options.showStatus, "status", false, "Show formatting status")
 	flags.BoolVar(&options.stopFormat, "stop", false, "Stop formatting progress")
+	flags.BoolVar(&options.debug, "debug", false, "Debug formatting progress")
+	flags.BoolVar(&options.cleanFormat, "clean", false, "Clean the Container")
```
Comment (caoxianfei1) [inline cli/command/format.go:126]: ditto

File: cli/command/format.go [hunk 10]
```diff
@@ -93,25 +135,36 @@ func genFormatPlaybook(curveadm *cli.CurveAdm,
 		return nil, errno.ERR_NO_DISK_FOR_FORMATTING
 	}
 
-	if options.showStatus && options.stopFormat {
-		return nil, errno.ERR_UNSUPPORT_CONFIGURE_VALUE_TYPE
-	}
+	showStatus  := options.showStatus
+	stopFormat  := options.stopFormat
+	debug       := options.debug
+	cleanFormat := options.cleanFormat
```
Comment (caoxianfei1) [inline cli/command/format.go:141]: ditto

**Comment by caoxianfei1** [comment]
lgtm

File: cli/command/format.go (hunk 8 above)
Comment (Wine93) [reply cli/command/format.go:79]: > clean instead of cleanFormat is better?

Agree!

File: cli/command/format.go [hunk 11]
```diff
@@ -41,7 +41,16 @@ const (
...
+  $ curveadm format --stop   -f /path/to/format.yaml  # Stop formatting progress
+  $ curveadm format --debug  -f /path/to/format.yaml  # Format chunkfile with debug mode
+  $ curveadm format --clean  -f /path/to/format.yaml  # clean the container left by debug mode`
+)
+
+const (
+	FORMAT_CHUNKFILE_POOL = playbook.FORMAT_CHUNKFILE_POOL
+	GET_FORMAT_STATUS     = playbook.GET_FORMAT_STATUS
+	STOP_FORMAT           = playbook.STOP_FORMAT
+	CLEAN_FORMAT          = playbook.CLEAN_FORMAT
```
Comment (caoxianfei1) [inline cli/command/format.go:53]: delete it if not used.

**Comment by caoxianfei1** [comment]
lgtm.

**Review by caoxianfei1 — APPROVED** [review]

**Review by Wine93 — APPROVED** [review]
LGTM!

**Pull request merged**